# How to Generate Implicit Judgment Lists from User Click Data

This notebook outlines a step by step process for generating a judgment list that can be used to calculate search engine relevance metrics like nDCG. A judgment list is a colleciton of query/document pairs given a relevance rating label. For example, for the query "red car", the document named "Lightning McQueen is a Red Racing Car" might be rated as highly relevant, or 3 on a 0-3 scale.

Once a judgment list is created, it can be used to calculate nDCG (and other relevance metrics) in a data analytics pipeline, or an opensource tool like Quepid.

### Notebook Outline

1. Install & Import Dependencies
2. Load the Data
3. Calculate Simple Click Through Rate (CTR)
4. Overcome Position Bias with Simplified Dynamic Bayesian Network
5. Overcome Confidence Bias with Beta Prior
6. Format the Judgment List

This whole process pulls heavily from Chapter 11 of the book ["AI-Powered Search"](https://www.manning.com/books/ai-powered-search) by Trey Grainger, Doug Turnbull, Max Irwin. For a deeper explanation of the process, and information for how to leverage a judgment list for AI features like Learn to Rank, please support the authors and buy the book!

# Install & Import Dependencies
All you'll need for this notebook is python (of course), and pandas to manipulate the data. You can provide your own data, or choose to load our full or small dataset.

In [ ]:
%pip install pandas
import pandas as pd

# Load the Data

The sample data is stored in a compressed pickle file which can be read directly into a pandas dataframe.

While the compressed file `implicit_judgments_data.pkl` is only a couple gigabytes, the original csv was over 24GB. The dataframe ends up being a little over 31GB in memory with the index. In other words, it could take a few minutes to load the df.

The `implicit_judgments_data_small.pkl` file is a smaller dataset that loads much quicker (a few seconds) and might be better for demo purposes. The original csv was only 500mb, and the dataframe is a little over 700mb when loaded in memory.

In [66]:
# uncomment the dataset you want to use, or load your own!

# dataset = 'implicit_judgments_data_small.pkl'
# dataset = 'implicit_judgments_data.pkl'

# load the dataset
df = pd.read_pickle(dataset, compression='bz2')

### Inspect the dataframe

In [199]:
df[:10]

session_id  user_id  query_id    doc_id  position  clicked
0    34573630       15  10509813  34175267         1        1
1    34573630       15  10509813  34171511         2        0
2    34573630       15  10509813  35444452         3        0
3    34573630       15  10509813  15370141         4        0
4    34573630       15  10509813  31342884         5        0
5    34573630       15  10509813  43630531         6        0
6    34573630       15  10509813  26065978         7        0
7    34573630       15  10509813  29902424         8        0
8    34573630       15  10509813  39016998         9        0
9    34573630       15  10509813  62861215        10        0

The first column on the left is the pandas dataframe built-in index (no label). Next we have a column for `session_id`, `user_id`, `query_id`, `doc_id`, `position`, and `clicked`. `session_id` refers to a single search session by a user (`user_id`), corresponding to a `query_id` (the identifier given to a query, aka search term), and the resulting documents (identified by `doc_id`). For a given search session we can see the given position of the result documents in the `position` column, and whether the document link was clicked or not (`clicked = 1` for clicked, `0` for not clicked).

In the `implicit_judgments_data_small` dataset, we can see the first search session has `session_id` `3457630`, made by user `15`. The `query_id` is `10509813` and we can see there were 10 documents returned, but only the first document in the result set was clicked.

> Note: the sample data only contains `query_id`, but ideally, your end judgment list would use the actual query string.

# Calculate Simple Click Through Rate (CTR)

To calculate CTR, we'll count all the clicks for a query that a document recieved, and divide it by the total number of search sessions where that query/doc pair occurred.

In [200]:
def calculate_ctr(sessions):
  num_clicks = sessions.groupby(["query_id", "doc_id"])["clicked"].sum()
  num_sessions = sessions.groupby(["query_id", "doc_id"])["session_id"].nunique()
  ctr = num_clicks / num_sessions
  return ctr

ctr = calculate_ctr(df)
ctr.sort_values(ascending=False)[:50]

query_id  doc_id  
2452707   58541183    2.0
7795064   42876266    2.0
15762397  56197529    2.0
4915918   1882964     2.0
669859    1235044     2.0
11979214  41935282    2.0
3507033   27638557    2.0
2963766   60197483    2.0
2099316   41490361    2.0
9928578   60635459    2.0
104584    70496331    2.0
11131380  7238661     2.0
20303545  16842490    2.0
104584    15660452    2.0
          11084634    2.0
669859    19537820    2.0
4100451   34033472    2.0
17185468  28830094    2.0
739545    10058203    2.0
21235243  35794486    2.0
609346    37038217    2.0
17185468  3367745     2.0
19413606  66134228    2.0
2704555   53914790    1.5
5302328   3272635     1.0
5302330   18618276    1.0
12048414  8829567     1.0
2913884   7210064     1.0
          26469811    1.0
6844000   26072969    1.0
5302330   39793713    1.0
2913884   46152681    1.0
8883434   52278349    1.0
14305044  63493627    1.0
15293149  30760668    1.0
14305044  53608280    1.0
15293149  15378814    1.0
          13713521 

# Overcome Position Bias with Simplified Dynamic Bayesian Network

CTR can be victim of biases in users and our application. One bias is Position Bias. Simply put, Position Bias occurs because a user is more likely to click on results towards the top of the search result page regardless of how relevant the result actually is.

One way to combat this is using a Simplified Dynamic Bayesian Network, which uses the idea of "examines" to essentially reward clicks that occur lower in the result list.

### Add examines to the dataframe
Every document above the last clicked result is thought to have been "examined", or observed/scanned by the user.


In [201]:
def add_examines(sessions):
  # filter to only clicked events
  clicked_sessions = sessions[sessions["clicked"] == 1]
  # Group by 'session_id' and 'query_id' and get the maximum 'position' for each session/query
  last_click_per_session = clicked_sessions.groupby(["session_id", "query_id"])["position"].max()

  # Reset index to make it easier to map
  last_click_per_session = last_click_per_session.reset_index()
  last_click_per_session = last_click_per_session.rename(columns={"position": "last_click_position"})

  # Merge the last_click_per_session with the original sessions DataFrame
  sessions = pd.merge(sessions, last_click_per_session, on=["session_id", "query_id"], how="left")
    
  # fill NaN last clicks (sessions with no clicks) with 0 value
  sessions['last_click_position'] = sessions["last_click_position"].fillna(0)
  
  # add the examined column, set to true if the position is less than or equal to the last clicked position
  sessions["examined"] = sessions["position"] <= sessions["last_click_position"] 
  return sessions

df_with_examines = add_examines(df)
df_with_examines

session_id  user_id  query_id    doc_id  position  clicked  \
0           34573630       15  10509813  34175267         1        1   
1           34573630       15  10509813  34171511         2        0   
2           34573630       15  10509813  35444452         3        0   
3           34573630       15  10509813  15370141         4        0   
4           34573630       15  10509813  31342884         5        0   
...              ...      ...       ...       ...       ...      ...   
15430955    35970414  5794351  11695257  61490336         6        0   
15430956    35970414  5794351  11695257  11153727         7        0   
15430957    35970414  5794351  11695257  31693680         8        0   
15430958    35970414  5794351  11695257  31603391         9        0   
15430959    35970414  5794351  11695257  61963518        10        0   

          last_click_position  examined  
0                         1.0      True  
1                         1.0     False  
2                         1.0     False  
3                         1.0     False  
4                         1.0     False  
...                       ...       ...  
15430955                  0.0     False  
15430956                  0.0     False  
15430957                  0.0     False  
15430958                  0.0     False  
15430959                  0.0     False  

[15430960 rows x 8 columns]

In [214]:
# for every query/doc pair, sum the clicks and examines for the doc in that query

def calculate_clicked_examined(sessions):
  return sessions[sessions["examined"]].groupby(["query_id","doc_id"])[["clicked", "examined"]].sum()

calculate_clicked_examined(df_with_examines)

clicked  examined
query_id doc_id                     
250      56587578        1         1
517      70880027        1         1
622      13698417        1         1
         42550710        0         1
955      15534388        1         1
...                    ...       ...
21432624 64097494        1         1
         65801311        0         1
21433077 41732620        1         2
21433466 26188386        1         1
         40568812        0         1

[978121 rows x 2 columns]

### Calculate Grade
Here we divide the total number of clicks by total number of examines for query/doc pair in all of our sessions.

In [215]:
def calculate_grade(sessions):
  sessions = calculate_clicked_examined(sessions)
  sessions["grade"] = sessions["clicked"] / sessions["examined"]
  return sessions

search_df = calculate_grade(df_with_examines)
# flatten the df multi-index so all columns are same level
search_df.reset_index(inplace=True)
search_df[search_df['grade'] > 0][:50].sort_values(by='grade', ascending=True)


query_id    doc_id  clicked  examined  grade
14       1343  15131650        1         2    0.5
27       1437  64481928        1         2    0.5
34       1791   2624065        1         2    0.5
87       6522  44239139        1         2    0.5
18       1343  67976152        1         2    0.5
12       1343    406395        1         2    0.5
13       1343  13171393        1         2    0.5
17       1343  55674579        1         2    0.5
73       5882  69886391        1         1    1.0
0         250  56587578        1         1    1.0
68       5722  39421635        1         1    1.0
67       5440  64824466        1         1    1.0
66       5440  40442197        1         1    1.0
64       5440  21067238        1         1    1.0
63       5440  12502637        1         1    1.0
74       6018   4423788        1         1    1.0
77       6381  17337894        1         1    1.0
85       6415  58045647        1         1    1.0
83       6415  44526278        1         1    1.0
61       4807  42179009        1         1    1.0
89       6823  13075839        1         1    1.0
90       7086  54621153        1         1    1.0
93       7454  29867818        1         1    1.0
94       7454  30576311        1         1    1.0
96       7468  14784360        1         1    1.0
100      7468  62047192        1         1    1.0
81       6415  38825408        1         1    1.0
60       4488  46996517        1         1    1.0
47       3691  12535128        1         1    1.0
53       3988  39028421        1         1    1.0
1         517  70880027        1         1    1.0
2         622  13698417        1         1    1.0
4         955  15534388        1         1    1.0
10       1265  54575630        1         1    1.0
11       1265  56593902        1         1    1.0
20       1344  70648215        1         1    1.0
21       1363   3145914        1         1    1.0
22       1363   6103078        1         1    1.0
56       4343  34432714        1         1    1.0
29       1459  19609656        1         1    1.0
31       1750  18071766        1         1    1.0
33       1750  38835181        1         1    1.0
35       1806  12257939        1         1    1.0
36       2967  44137761        1         1    1.0
41       3024  38347446        1         1    1.0
44       3414   6270878        1         1    1.0
102      7891  58992691        1         1    1.0
51       3691  66646896        1         1    1.0
30       1750   8334432        1         1    1.0
103      8020  59789057        1         1    1.0

# Overcome Confidence Bias with Beta Distribution
Should a document with 1 examine and 1 click have a grade of 1 (100%)? Probably not -- there's not enough information to be able to tell if that was just a random click. Beta Distribution is one way we can estimate what a typical mean grade would be, and then create a probability distribution that increases slightly for the few occurences of the click/examine events.

In other words, it assumes what a typical grade should be and then acts as if 1 examine and 1 click are just one drop in the bucket of all the clicks and examines that led to the typical grade. The 1 examine and 1 click will "pull" the distribution a little to the "higher" probablity according to a weight we'll assign.

We'll assign our typical grade (`prior_grade`) the mean grade of all our events. We'll assign it a weight of `100`. These values can be played around with. The more the weight, the more the probability distribution will be affected by click events.

### Assign prior_a and prior_b values

In [216]:
def calculate_prior(sessions, prior_grade, prior_weight):
  sessions["prior_a"] = prior_grade * prior_weight
  sessions["prior_b"] = (1 - prior_grade) * prior_weight
  return sessions

median_grade = search_df['grade'].median()

prior_grade = median_grade
prior_weight = 100
calculate_prior(search_df, prior_grade, prior_weight)

query_id    doc_id  clicked  examined  grade    prior_a    prior_b
0            250  56587578        1         1    1.0  16.666667  83.333333
1            517  70880027        1         1    1.0  16.666667  83.333333
2            622  13698417        1         1    1.0  16.666667  83.333333
3            622  42550710        0         1    0.0  16.666667  83.333333
4            955  15534388        1         1    1.0  16.666667  83.333333
...          ...       ...      ...       ...    ...        ...        ...
978116  21432624  64097494        1         1    1.0  16.666667  83.333333
978117  21432624  65801311        0         1    0.0  16.666667  83.333333
978118  21433077  41732620        1         2    0.5  16.666667  83.333333
978119  21433466  26188386        1         1    1.0  16.666667  83.333333
978120  21433466  40568812        0         1    0.0  16.666667  83.333333

[978121 rows x 7 columns]

### Calculate posterior_a, posterior_b values, and final beta_grade value

In [217]:
def calculate_sdbn(sessions, prior_grade=0.3, prior_weight=100):
  sessions = calculate_prior(sessions, prior_grade, prior_weight)
  sessions["posterior_a"] = (sessions["prior_a"] + sessions["clicked"])
  sessions["posterior_b"] = (sessions["prior_b"] + sessions["examined"] - sessions["clicked"])
  sessions["beta_grade"] = (sessions["posterior_a"] / (sessions["posterior_a"] + sessions["posterior_b"]))
  return sessions.sort_values("beta_grade", ascending=False)

sessions = search_df
calculate_sdbn(sessions)

query_id    doc_id  clicked  examined     grade  prior_a  prior_b  \
915961  20100007  46063140      340       408  0.833333     30.0     70.0   
150830   4605457  39061378      442       570  0.775439     30.0     70.0   
132876   4102451  34175267     1247      1866  0.668274     30.0     70.0   
489988  11483526  30581891      237       317  0.747634     30.0     70.0   
797870  17670982  28406892      131       157  0.834395     30.0     70.0   
...          ...       ...      ...       ...       ...      ...      ...   
149297   4575281  39002266        1        38  0.026316     30.0     70.0   
293974   7444613   6746234        3        49  0.061224     30.0     70.0   
149299   4575281  48039998        0        36  0.000000     30.0     70.0   
485541  11401789  63399546       11        86  0.127907     30.0     70.0   
150827   4605457  29220341        5        60  0.083333     30.0     70.0   

        posterior_a  posterior_b  beta_grade  
915961        370.0        138.0    0.728346  
150830        472.0        198.0    0.704478  
132876       1277.0        689.0    0.649542  
489988        267.0        150.0    0.640288  
797870        161.0         96.0    0.626459  
...             ...          ...         ...  
149297         31.0        107.0    0.224638  
293974         33.0        116.0    0.221477  
149299         30.0        106.0    0.220588  
485541         41.0        145.0    0.220430  
150827         35.0        125.0    0.218750  

[978121 rows x 10 columns]

# Format the Judgment List
We now have all the data we need, but we can clean it up a bit to make calculating nDCG a bit easier. First we will quantize our `beta_grade` probabilistic float values into four quantiles, which will get assigned a value from 0-3 in the new `grade` column.

Then we'll drop the columns we don't need, et voila. Implicit Judgment List.

### Quantize

In [194]:
search_df['grade'] = pd.cut(search_df['beta_grade'], bins=4, labels=False)
search_df.sort_values(by='grade', ascending=False)[:50]

query_id    doc_id  clicked  examined  grade  prior_a  prior_b  \
328250   8117795   8539216      115       138      3     30.0     70.0   
132876   4102451  34175267     1247      1866      3     30.0     70.0   
489988  11483526  30581891      237       317      3     30.0     70.0   
110065   3446059  26785689      180       243      3     30.0     70.0   
453269  10751928  51205217      202       271      3     30.0     70.0   
150830   4605457  39061378      442       570      3     30.0     70.0   
441089  10509813  34175267      227       327      3     30.0     70.0   
797870  17670982  28406892      131       157      3     30.0     70.0   
603726  13838503  31747249      110       125      3     30.0     70.0   
915923  20098967  46063140      114       131      3     30.0     70.0   
915961  20100007  46063140      340       408      3     30.0     70.0   
409330   9819082  50172603      123       203      2     30.0     70.0   
123121   3812356  31039976      114       167      2     30.0     70.0   
519602  12087407  53028782       50        60      2     30.0     70.0   
328058   8107157   8538730       52        64      2     30.0     70.0   
409136   9814521  50172603      117       182      2     30.0     70.0   
649313  14745379  53028782      117       161      2     30.0     70.0   
188919   5353237  40242351      132       172      2     30.0     70.0   
513677  11958661  63254471       59        82      2     30.0     70.0   
471713  11113019   9916808      125       164      2     30.0     70.0   
941914  20649304  41426231      106       137      2     30.0     70.0   
161865   4831705  37492192       85       128      2     30.0     70.0   
440927  10506605  42239078       48        64      2     30.0     70.0   
925029  20303941   9158898       54        75      2     30.0     70.0   
480605  11297354  53028782       43        51      2     30.0     70.0   
438046  10453943  15569609       58        77      2     30.0     70.0   
193865   5439298  40519741       56        70      2     30.0     70.0   
483800  11364848  53170009       73        97      2     30.0     70.0   
621209  14200002  47234398      210       313      2     30.0     70.0   
875721  19236076  65506930       74       114      2     30.0     70.0   
342798   8401734  27160700       76       113      2     30.0     70.0   
348352   8516796  58110589       81       115      2     30.0     70.0   
463160  10954331  51564488       74        96      2     30.0     70.0   
816714  18103292  63055573       82       133      2     30.0     70.0   
170886   4989763  63249531      121       152      2     30.0     70.0   
189498   5366856  40266487       54        73      2     30.0     70.0   
766188  17030180  48046941       59        80      2     30.0     70.0   
117934   3666060  63478150       77       110      2     30.0     70.0   
826499  18281156  63501165      130       186      2     30.0     70.0   
649594  14753069  17286189       55        76      2     30.0     70.0   
189509   5366936  40266487       65        77      2     30.0     70.0   
176115   5090291  37216570       84       124      2     30.0     70.0   
967486  21208690  44469547       86       126      2     30.0     70.0   
235871   6326563  67795671       78       102      2     30.0     70.0   
188944   5353542  40242351       48        64      2     30.0     70.0   
884254  19413011  69827886       53        71      2     30.0     70.0   
45795    1435646  34175267       78       103      2     30.0     70.0   
59477    1856912  31747249       88       114      2     30.0     70.0   
915432  20093215  46055958       64        93      2     30.0     70.0   
513722  11959283  63254471       96       132      2     30.0     70.0   

        posterior_a  posterior_b  beta_grade  
328250        145.0         93.0    0.609244  
132876       1277.0        689.0    0.649542  
489988        267.0        150.0    0.640288  
110065        210.0        133.0    0.612245 

### Format and drop unused columns

> Again, note that in a real judgment list -- we'd want the _actual_ query strings instead of a `query_id`.

In [219]:
columns_to_drop = ['clicked', 'examined', 'prior_a', 'prior_b', 'posterior_a', 'posterior_b', 'beta_grade']
final_judgment_list = search_df.drop(columns_to_drop, axis='columns')
final_judgment_list.set_index("query_id", inplace=True)
final_judgment_list.sort_values(by='grade', ascending=False)[:50]

doc_id  grade
query_id                 
250       56587578    1.0
11710437  26135987    1.0
11711488  50559245    1.0
11711438  62274753    1.0
11711438  58375232    1.0
11711438  57372829    1.0
11711438  28074867    1.0
11711438  23398553    1.0
11711431  37564552    1.0
11711430  62832733    1.0
11711430  61885811    1.0
11711430  48512988    1.0
11711430  43399910    1.0
11711430  30644224    1.0
11711428  62274427    1.0
11711428  43399919    1.0
11711428   3439797    1.0
11711427  63562866    1.0
11711427   1108224    1.0
11711337  26137618    1.0
11711309  59542133    1.0
11711309   5360686    1.0
11711201  24144949    1.0
11711201  18194921    1.0
11711201  18194893    1.0
11710667  25447944    1.0
11710603  67563433    1.0
11710603  44026441    1.0
11710570  54322099    1.0
11710551  54776136    1.0
11710515  55034389    1.0
11711536  37607269    1.0
11711536  49073724    1.0
11711546  26136062    1.0
11712989  15451001    1.0
11714063  60771082    1.0
11714063  41223119    1.0
11714063  30160365    1.0
11713987  62757092    1.0
11713987  54330494    1.0
11713369  51898389    1.0
11713369  49957485    1.0
11713369  46065825    1.0
11713183  19659970    1.0
11712991  54070477    1.0
11712991  54070346    1.0
11712989  54070477    1.0
11712989  54070346    1.0
11712936  54070346    1.0
11711606  65087497    1.0

### Export to CSV

In [220]:
final_judgment_list.to_csv("final_judgment_list.csv")